### PR TITLE
Force trusted setup if circuits have been modified

### LINKED ISSUE #989: Force trusted setup is circuits have been modified
Currently, trusted setup is only done if either keys are not found or it its forced. However, it circuits have been modified and no trusted setup is forced the transaction will fail the verification.  This issues proposes a mechanism to overcome this. It will provide a better usability so that circuits that are modified do always go through a trusted setup.

The mechanism could be something like:
- In worker we have circuits/deposit.zok and we have output/deposit/deposit_out.
- In the output/  we should also store a hash.txt 
- Then we can always hash(deposit.zok)  and see if it matches the value in hash.txt if not (or hash.txt doesnt exisit) then we must re-do the setup
```
for (let i = 0; i < vks.length; i++) {
    const circuit = circuitsToSetup[i];
    ----> const circuitHash = generateHash(circuit); // computes md5 hash of .zok circuit source file
    ----> const prestoredCircuitHash = loadCircuitHash(i); // loads prestored hash of circuit i from file in output/
    if (!vks[i] || config.ALWAYS_DO_TRUSTED_SETUP || circuitHash !== prestoredCircuitHash ) {
      // we don't have an existing vk so let's generate one
      try {
        logger.info(
          `no existing verification key. Fear not, I will make a new one: calling generate keys on ${circuit}`,
        );
       
        const res2 = await axios.post(
          `${config.PROTOCOL}${config.ZOKRATES_WORKER_HOST}/generate-keys`,
          {
            filepath: circuit,
            curve: config.CURVE,
            provingScheme: config.PROVING_SCHEME,
            backend: config.BACKEND,
          },
        );
        vks[i] = res2.data.vk;
        -----> storeCircuitHash(i, circuitHash);   ------> update circuit hash in output/
      } catch (err) {
        logger.error(err);
      }
    } else logger.info(`${circuit} verification key exists: trusted setup skipped`);
  }
```

### PR DESCRIPTION
**Pull request by eugypalu**
## What does this implement/fix? Explain your changes.
The goal of this pr is to check if there have been any changes to the circuits and if so, go to force a trusted setup. 
To do this is kept within the `/output/circuithash.txt` (in the worker), the list of all circuits with their hash. 
So it checks if the hash of the current circuit differs, or is not in the file, and in case a trusted setup is forced

## Does this close any currently open issues?
Resolves #989

### CODE REVIEW

File: config/default.js [hunk 1]
```diff
@@ -31,7 +31,6 @@ module.exports = {
   TXHASH_TREE_HASH_TYPE: 'keccak256',
   STATE_GENESIS_BLOCK: process.env.STATE_GENESIS_BLOCK,
   CIRCUITS_HOME: process.env.CIRCUITS_HOME || '/app/circuits/',
-  ALWAYS_DO_TRUSTED_SETUP: process.env.ALWAYS_DO_TRUSTED_SETUP || false,
```
Comment (druiz0992) [inline config/default.js:34]: We still want to preserve the ability to force a trusted setup

File: nightfall-deployer/src/circuit-setup.mjs [hunk 2]
```diff
@@ -92,8 +93,18 @@ async function setupCircuits() {
   for (let i = 0; i < vks.length; i++) {
     const circuit = circuitsToSetup[i];
 
-    if (!vks[i] || config.ALWAYS_DO_TRUSTED_SETUP) {
-      // we don't have an existing vk so let's generate one
+    const fileBuffer = fs.readFileSync(`./circuits/${circuit}`);
+    const hcircuit = crypto.createHash('md5').update(fileBuffer).digest('hex');
+
+    const checkHash = await axios.post(
```
Comment (druiz0992) [inline nightfall-deployer/src/circuit-setup.mjs:99]: There needs to be a way for force a trusted setup. Also, what happens if ther eis not vk[i] defined?

**Review by druiz0992 — CHANGES_REQUESTED** [review]
Some additional questions:
- What happens if there are new circuits where hash doesnt match?

File: config/default.js (hunk 1 above)
Comment (eugypalu) [reply config/default.js]: Done, reverted this change

File: nightfall-deployer/src/circuit-setup.mjs (hunk 2 above)
Comment (eugypalu) [reply nightfall-deployer/src/circuit-setup.mjs:99]: I restored the ability to force trusted setup. 
from the tests I did in the case vk[i] is undefined there was always also a difference on the hash, but for safety I added the vk[i] check, if undefined trusted setup is forced

**Comment by eugypalu** [comment]
> 

in case new circuits were added: 

-if the hash is present in the file but does not match the one just calculated, the hash contained in the file is updated and the trusted setup is generated.

-if, on the other hand, the hash is not present in the file, it is added to the file and the trusted setup is generated

File: nightfall-deployer/src/circuit-setup.mjs (hunk 2 above)
Comment (druiz0992) [reply nightfall-deployer/src/circuit-setup.mjs:99]: since you already have tests, can you also include those? or were they manual?

**Review by RogerTaule — APPROVED** [review]

File: nightfall-deployer/src/circuit-setup.mjs (hunk 2 above)
Comment (eugypalu) [reply nightfall-deployer/src/circuit-setup.mjs:99]: no, sorry but they were all manual tests

**Review by RogerTaule — COMMENTED** [review]
Thanks for adding the getCircuitHash enpdoint :)

**Pull request merged**